### PR TITLE
Make sure organization groups can not be managed but when managing an organization

### LINKED ISSUE #29431: Make sure organization groups can not be managed but when managing an organization
### Description

A follow-up on https://github.com/keycloak/keycloak/issues/28734 so that we can:

* Leverage provider events to run validations when creating/updating/removing groups
* Cover not only managing groups through the Admin REST API but also when managing groups from other place like from mappers or when using the group internally
* Make sure the validation happens in a single place and within the organization provider rather than spread everywhere in the code base

### Discussion

_No response_

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by pedroigor**
Closes #29431

* Leveraging provider events to introduce validations when managing organization groups
* Introduces additional events to groups that might be also useful for regular realms
* Keeping the validation within the organization provider rather than spread in multiple places

### CODE REVIEW

File: services/src/main/java/org/keycloak/services/resources/admin/GroupsResource.java [hunk 1]
```diff
@@ -124,11 +121,14 @@ public Stream<GroupRepresentation> getGroups(@QueryParam("search") String search
     @Path("{group-id}")
     public GroupResource getGroupById(@PathParam("group-id") String id) {
         GroupModel group = realm.getGroupById(id);
+
         if (group == null) {
             throw new NotFoundException("Could not find group by id");
         }
 
-        checkForOrgRelatedGroupModel(session, group);
+        if (!Organizations.canManageOrganizationGroup(session, group)) {
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/services/resources/admin/GroupsResource.java:129]: We can safely remove the validation from here and have the organization provider as the source of truth for validations, so a single place where it happens.

From an UI perspective, we could probably show organizational groups disabled based on the `kc.org` attribute.

File: model/jpa/src/main/java/org/keycloak/organization/jpa/JpaOrganizationProvider.java [hunk 2]
```diff
@@ -76,21 +78,22 @@ public OrganizationModel create(String name, Set<String> domains) {
...
+        OrganizationAdapter adapter = new OrganizationAdapter(realm, this);
 
-        GroupModel group = createOrganizationGroup(entity.getId());
-
-        entity.setGroupId(group.getId());
-        entity.setRealmId(realm.getId());
-        entity.setName(name);
-        entity.setEnabled(true);
+        try {
+            session.setAttribute(OrganizationModel.class.getName(), adapter);
```
Comment (pedroigor) [inline model/jpa/src/main/java/org/keycloak/organization/jpa/JpaOrganizationProvider.java:84]: We need this because the validations happen based on whether the `OrganizationModel` is not available in the session. If not, it means we are running outside the provider boundaries so we should fail any attempt to update/remove the group.

I'm also not very happy about this boilerplate code but otherwise, we will need to keep the previous changes where the validation happens in multiple places, something I think is not good from a maintenance perspective.

Also, the previous changes are missing places where groups are managed other than the Admin REST API. For instance, when running mappers or managing groups internally.

I don't have an alternative to achieve the same goal (have validations in a single place) so please let me know if you other suggestion.

File: model/jpa/src/main/java/org/keycloak/organization/jpa/JpaOrganizationProvider.java [hunk 3]
```diff
@@ -297,10 +317,22 @@ public boolean removeMember(OrganizationModel organization, UserModel member) {
         if (isManagedMember(organization, member)) {
             userProvider.removeUser(realm, member);
         } else {
-            List<String> organizations = member.getAttributes().get(ORGANIZATION_ATTRIBUTE);
-            organizations.remove(organization.getId());
-            member.setAttribute(ORGANIZATION_ATTRIBUTE, organizations);
-            member.leaveGroup(getOrganizationGroup(organization));
+            OrganizationModel current = (OrganizationModel) session.getAttribute(OrganizationModel.class.getName());
+
+            if (current == null) {
```
Comment (pedroigor) [inline model/jpa/src/main/java/org/keycloak/organization/jpa/JpaOrganizationProvider.java:322]: This check is because we don't want to remove the attribute if an organization has been previously set.

File: services/src/main/java/org/keycloak/organization/utils/Organizations.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import org.keycloak.common.Profile;
+import org.keycloak.common.Profile.Feature;
+import org.keycloak.models.GroupModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.OrganizationModel;
+import org.keycloak.utils.StringUtil;
+
+public class Organizations {
+
+    public static boolean canManageOrganizationGroup(KeycloakSession session, GroupModel group) {
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/organization/utils/Organizations.java:29]: This validation should happen even if organizations are disabled in the realm.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/organization/admin/OrganizationTest.java [hunk 5]
```diff
@@ -440,17 +447,68 @@ public void testManageOrgGroupsViaDifferentAPIs() {
...
-        } catch (ForbiddenException ex) {
+            fail("Expected BadRequestException");
+        } catch (BadRequestException ex) {
             // success
         }
+    }
 
-        try {
-            // cannot leave organization related group
-            testRealm().users().get(userRep.getId()).leaveGroup(orgGroupRep.getId());
```
Comment (martin-kanis) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/organization/admin/OrganizationTest.java:450]: Why was this removed from the test?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/organization/admin/OrganizationTest.java [hunk 6]
```diff
@@ -414,23 +422,22 @@ public void testManageOrgGroupsViaDifferentAPIs() {
...
             // cannot add organization related group as a default group
             testRealm().addDefaultGroup(orgGroupRep.getId());
-            fail("Expected ForbiddenException");
-        } catch (ForbiddenException ex) {
-            // success
-        }
-
-        try {
-            // cannot remove organization related group as a default group
-            testRealm().removeDefaultGroup(orgGroupRep.getId());
```
Comment (martin-kanis) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/organization/admin/OrganizationTest.java:431]: Why was this removed from the test?

File: services/src/main/java/org/keycloak/services/resources/admin/GroupResource.java [hunk 7]
```diff
@@ -118,11 +118,14 @@ public Response updateGroup(GroupRepresentation rep) {
         this.auth.groups().requireManage(group);
 
         String groupName = rep.getName();
+
         if (ObjectUtil.isBlank(groupName)) {
             throw ErrorResponse.error("Group name is missing", Response.Status.BAD_REQUEST);
         }
 
-        checkForOrgRelatedGroupRep(session, rep);
+        if (rep.getId() != null && !group.getId().equals(rep.getId())) {
```
Comment (martin-kanis) [inline services/src/main/java/org/keycloak/services/resources/admin/GroupResource.java:126]: Why this change in Group API is necessary?

File: model/jpa/src/main/java/org/keycloak/organization/jpa/JpaOrganizationProvider.java [hunk 8]
```diff
@@ -76,21 +78,22 @@ public OrganizationModel create(String name, Set<String> domains) {
...
+            adapter.setName(name);
+            adapter.setEnabled(true);
 
-        OrganizationAdapter adapter = new OrganizationAdapter(realm, entity, this);
+            em.persist(adapter.getEntity());
 
-        adapter.setDomains(domains.stream().map(OrganizationDomainModel::new).collect(Collectors.toSet()));
+            adapter.setDomains(domains.stream().map(OrganizationDomainModel::new).collect(Collectors.toSet()));
+        } finally {
+            session.removeAttribute(OrganizationModel.class.getName());
```
Comment (martin-kanis) [inline model/jpa/src/main/java/org/keycloak/organization/jpa/JpaOrganizationProvider.java:95]: Do we need to remove the attribute from the session? Can we add it in OrganizationResource and in the constructor of the JpaOrganizationProvider and leave it there for the lifespan of the session?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/organization/admin/OrganizationTest.java (hunk 6 above)
Comment (pedroigor) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/organization/admin/OrganizationTest.java:431]: Because you can't add a default group.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/organization/admin/OrganizationTest.java (hunk 5 above)
Comment (pedroigor) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/organization/admin/OrganizationTest.java:450]: Because you can't join a group.

File: services/src/main/java/org/keycloak/services/resources/admin/GroupResource.java (hunk 7 above)
Comment (pedroigor) [reply services/src/main/java/org/keycloak/services/resources/admin/GroupResource.java:126]: Because otherwise, you are accepting a rep for a group other than the one you are actually managing from the API.

File: model/jpa/src/main/java/org/keycloak/organization/jpa/JpaOrganizationProvider.java (hunk 8 above)
Comment (pedroigor) [reply model/jpa/src/main/java/org/keycloak/organization/jpa/JpaOrganizationProvider.java:95]: Yes, it does not make sense to keep it there after this method is called.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/organization/admin/OrganizationTest.java (hunk 5 above)
Comment (pedroigor) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/organization/admin/OrganizationTest.java:450]: I added this one back.

**Review by sguilhen — APPROVED** [review]

**Review by martin-kanis — APPROVED** [review]
LGTM

**Review by ssilvert — APPROVED** [review]
Approving based on the other two reviews.
